File: II-146-3/BridgeCrossing.cs
using System;
using System.Collections;
using System.Text;
using System.Text.RegularExpressions;

public class BridgeCrossing
{
	private Hashtable l = new Hashtable( 1000 );
	private Hashtable r = new Hashtable( 1000 );

	private int goRight( int[] left, int[] right )
	{
		if ( left.Length == 2 )
		{
			return left[0] > left[1] ? left[0] : left[1];
		}
		if ( l.Contains( left ) )
		{
			return (int) l[left];
		}
		int min = 6000;
		for( int i = 0; i < left.Length - 1; i++ )
		{
			for( int j = i + 1; j < left.Length; j++ )
			{
				int[] a = new int[left.Length - 2];
				int[] b = new int[right.Length + 2];
				int k = 0;
				for( int x = 0; x < left.Length; x++ )
				{
					if ( x != i && x != j )
					{
						a[k++] = left[x];
					}
				}				
				k = 0;
				for( int x = 0; x < right.Length; x++ )
				{
					b[k++] = right[x];
				}
				b[k++] = left[i];
				b[k++] = left[j];
				int result = goLeft( a, b ) + (left[i] > left[j] ? left[i] : left[j]);
				min = min < result ? min : result;
			}
		}
		l.Add( left, min );
		return min;
	}

	private int goLeft( int[] left, int[] right )
	{
		if ( r.Contains( right ) )
		{
			return (int) r[left];
		}
		int min = 6000;
		for( int i = 0; i < right.Length; i++ )
		{
			int[] a = new int[left.Length + 1];
			int[] b = new int[right.Length - 1];
			int k = 0;
			for( int x = 0; x < left.Length; x++ )
			{
				a[k++] = left[x];
			}
			a[k++] = right[i];
			k = 0;
			for( int x = 0; x < right.Length; x++ )
			{
				if ( x != i )
				{
					b[k++] = right[x];
				}				
			}
			int result = goRight( a, b ) + right[i];
			min = min < result ? min : result;
		}
		r.Add( right, min );
		return min;
	}

    public int minTime( int[] times )
    {
        if ( times.Length == 1 ) return times[0];
		return goRight( times, new int[] {} );
    }

// BEGIN CUT HERE
  public static void Main(string[] args)
  {
    //try
    {
		eq(1,(new BridgeCrossing()).minTime(new int[] { 1, 2, 5, 10 }),17);
		eq(2,(new BridgeCrossing()).minTime(new int[] { 1, 2, 3, 4, 5 }),16);
		eq(3,(new BridgeCrossing()).minTime(new int[] { 100 }),100);
		eq(4,(new BridgeCrossing()).minTime(new int[] { 1, 2, 3, 50, 99, 100 }),162);
    }
/*    catch(Exception e)
    {
	Console.WriteLine(e.Message);
    }*/
  }
  private static void eq(int n, int a, int b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, char a, char b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
  }
  private static void eq(int n, long a, long b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, bool a, bool b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, string a, string b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, int[] a, int[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, long[] a, long[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, string[] a, string[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void print(int a)
  {
    Console.Write(a+" ");
  }
  private static void print(string s)
  {
    Console.Write("\""+s+"\" ");
  }
  private static void print(int[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(long[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(string[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write("\""+rs[i]+"\"");
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void nl()
  {
    Console.WriteLine();
  }
// END CUT HERE
}

File: II-146-3/BridgeCrossing.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
#define all(v) (v).begin(), (v).end()
struct BridgeCrossing
{
	int minTime( vi times )
	{
		int n = times.size();
		if ( n == 1 ) return times[0];
		sort( all(times) );
		vi states( 1 << (n + 1), 5000 );
		queue<int> q; q.push( (1 << n) - 1 ); q.push( 0 ); q.push( 1 );
		while( !q.empty() )
		{
			int state = q.front(); q.pop();
			int time = q.front(); q.pop();
			int direction = q.front(); q.pop();
			if ( state )
				if ( direction )
				{
					for( int i = 0; i < n; ++i )
						if ( state & (1 << i) )
							for( int j = i + 1; j < n; ++j )
								if ( state & (1 << j) )
								{
									int newState = state - (1 << i) - (1 << j), newTime = time + times[j];
									if ( states[2 * newState + 1] > newTime )
									{
										states[2 * newState + 1] = newTime;
										q.push( newState ); q.push( newTime ); q.push( !direction );
									}
								}
				}
				else
				{
					for( int j = 0; j < n; ++j )
						if ( !(state & (1 << j)) )
						{
							int newState = state + (1 << j), newTime = time + times[j];
							if ( states[2 * newState] > newTime )
							{
								states[2 * newState] = newTime;
								q.push( newState ); q.push( newTime ); q.push( !direction );
							}
						}
				}
		}
		return states[1];
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vi> A0; vector<int> A1;
	{ /*Test 0*/int _times[] = 
	{ 1, 2, 5, 10 }
	; vi times(_times, _times + (sizeof(_times) / sizeof(_times[0]))); A0.push_back( times ); int result = 
	17
	; A1.push_back( result );  }
	{ /*Test 1*/int _times[] = 
	{ 1, 2, 3, 4, 5 }
	; vi times(_times, _times + (sizeof(_times) / sizeof(_times[0]))); A0.push_back( times ); int result = 
	16
	; A1.push_back( result );  }
	{ /*Test 2*/int _times[] = 
	{ 100 }
	; vi times(_times, _times + (sizeof(_times) / sizeof(_times[0]))); A0.push_back( times ); int result = 
	100
	; A1.push_back( result );  }
	{ /*Test 3*/int _times[] = 
	{ 1, 2, 3, 50, 99, 100 }
	; vi times(_times, _times + (sizeof(_times) / sizeof(_times[0]))); A0.push_back( times ); int result = 
	162
	; A1.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A1[i], minTime( A0[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	BridgeCrossing ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE
